mobilecoind/src/sync.rs

```diff
@@ -286,55 +294,108 @@ fn sync_thread_entry_point(
+    let parallel_results = (0..num_blocks_to_load)
+        .into_par_iter()
+        .map(
+            |worker_idx| -> Result<(Vec<UnspentTxOut>, Vec<KeyImage>), Error> {
+                // We don't expect to get Error::NotFound here, since we earlier tested
+                // ledger_db.num_blocks(). The other error cases are not common
+                // either.
+                let block_idx = monitor_data.next_block + worker_idx as u64;
+                let block_contents = ledger_db.get_block_contents(block_idx)?;
+
+                log::trace!(
+                    logger,
+                    "processing {} outputs and {} key images from block {} for monitor_id {}",
+                    block_contents.outputs.len(),
+                    block_contents.key_images.len(),
+                    block_idx,
+                    monitor_id,
+                );
+
+                // Match tx outs into UTXOs.
+                let utxos = match_tx_outs_into_utxos(
+                    mobilecoind_db,
+                    &block_contents.outputs,
+                    monitor_id,
+                    &monitor_data,
+                    logger,
+                )?;
+
+                Ok((utxos, block_contents.key_images))
+            },
+        )
+        .collect::<Vec<Result<(Vec<UnspentTxOut>, Vec<KeyImage>), Error>>>();
+
+    // For diagnostics, we want to keep track of which workers were successful.
+    // Usually it should be all of them, but if sometimes a worker in the middle
+    // fails and later ones don't, that will harm performance.
+    let worker_successes = parallel_results
+        .iter()
+        .map(|result| if result.is_ok() { 1 } else { 0 })
+        .collect::<Vec<usize>>();
+
+    // Now add everything to the database
+    for (worker_idx, result) in parallel_results.into_iter().enumerate() {
+        if let Err(err) = result.and_then(|(utxos, key_images)| {
+            // Update database.
+            mobilecoind_db.block_processed(
+                monitor_id,
+                monitor_data.next_block + worker_idx as u64,
+                &utxos,
+                &key_images,
+            )
+        }) {
+            // Unfortunately, we have to abandon any work that could have been accomplished
+            // successfully by the threads after this one. To track this,
+            // we'll log a warning about work being abandoned.
+            let abandoned_successes: usize = worker_successes.iter().skip(worker_idx + 1).sum();
+            if abandoned_successes > 0 {
+                log::warn!(logger, "Due to an error while parallel scanning, had to abandon {} successful block scanning results", abandoned_successes);
+            }
+            return Err(err);
+        }
     }
 
-    Ok(SyncMonitorOk::MoreBlocksPotentiallyAvailable)
+    Ok(if blocks_remaining_for_monitor == num_blocks_to_load {
+        SyncMonitorOk::NoMoreBlocks
+    } else {
+        SyncMonitorOk::MoreBlocksPotentiallyAvailable
+    })
```

Comment:
@eranrund I know we thought this if here would not be necessary, but I found I get a lot of test failures if I don't have this, because many of the tests call these functions and test for specific error codes that would have matched the earlier behavior. This if looks okay to me and it seems like the way to go.

Here are the errors I got without this:

```
failures:

---- sync::test::test_utxo_value_zero stdout ----
thread 'sync::test::test_utxo_value_zero' panicked at mobilecoind/src/sync.rs:904:9:
assertion `left == right` failed
  left: MoreBlocksPotentiallyAvailable
 right: NoMoreBlocks
stack backtrace:
   0: rust_begin_unwind
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/panicking.rs:270:5
   4: mc_mobilecoind::sync::test::__wrapped_test_utxo_value_zero
             at ./src/sync.rs:904:9
   5: mc_mobilecoind::sync::test::test_utxo_value_zero::{{closure}}
             at ./src/sync.rs:867:5
   6: slog_scope::scope
             at /tmp/mobilenode/.mob/cargo/registry/src/index.crates.io-6f17d22bba15001f/slog-scope-4.4.0/lib.rs:248:5
   7: mc_mobilecoind::sync::test::test_utxo_value_zero
             at ./src/sync.rs:867:5
   8: mc_mobilecoind::sync::test::test_utxo_value_zero::{{closure}}
             at ./src/sync.rs:867:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/ops/function.rs:250:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- sync::test::test_sync_monitor stdout ----
thread 'sync::test::test_sync_monitor' panicked at mobilecoind/src/sync.rs:611:9:
assertion `left == right` failed
  left: MoreBlocksPotentiallyAvailable
 right: NoMoreBlocks
stack backtrace:
   0: rust_begin_unwind
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/panicking.rs:270:5
   4: mc_mobilecoind::sync::test::__wrapped_test_sync_monitor
             at ./src/sync.rs:611:9
   5: mc_mobilecoind::sync::test::test_sync_monitor::{{closure}}
             at ./src/sync.rs:498:5
   6: slog_scope::scope
             at /tmp/mobilenode/.mob/cargo/registry/src/index.crates.io-6f17d22bba15001f/slog-scope-4.4.0/lib.rs:248:5
   7: mc_mobilecoind::sync::test::test_sync_monitor
             at ./src/sync.rs:498:5
   8: mc_mobilecoind::sync::test::test_sync_monitor::{{closure}}
             at ./src/sync.rs:498:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/ops/function.rs:250:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

---- sync::test::test_sync_monitor_with_random_recipients stdout ----
thread 'sync::test::test_sync_monitor_with_random_recipients' panicked at mobilecoind/src/sync.rs:794:9:
assertion `left == right` failed
  left: MoreBlocksPotentiallyAvailable
 right: NoMoreBlocks
stack backtrace:
   0: rust_begin_unwind
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/std/src/panicking.rs:597:5
   1: core::panicking::panic_fmt
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/panicking.rs:72:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/panicking.rs:270:5
   4: mc_mobilecoind::sync::test::__wrapped_test_sync_monitor_with_random_recipients
             at ./src/sync.rs:794:9
   5: mc_mobilecoind::sync::test::test_sync_monitor_with_random_recipients::{{closure}}
             at ./src/sync.rs:677:5
   6: slog_scope::scope
             at /tmp/mobilenode/.mob/cargo/registry/src/index.crates.io-6f17d22bba15001f/slog-scope-4.4.0/lib.rs:248:5
   7: mc_mobilecoind::sync::test::test_sync_monitor_with_random_recipients
             at ./src/sync.rs:677:5
   8: mc_mobilecoind::sync::test::test_sync_monitor_with_random_recipients::{{closure}}
             at ./src/sync.rs:677:5
   9: core::ops::function::FnOnce::call_once
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/ops/function.rs:250:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/ca62d2c445628587660ae48013f460b08b1f5552/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    sync::test::test_sync_monitor
    sync::test::test_sync_monitor_with_random_recipients
    sync::test::test_utxo_value_zero
```